---
title: "creigs-gwas-lecture6"
author: "cporras"
date: "2023-10-24"
output: html_document
---

# GWAS tutorial Lecture 6

Following tutorial at: https://www.r-bloggers.com/2017/10/genome-wide-association-studies-in-r/

Using tutorial github repo: 
https://github.com/monogenea/GWAStutorial

Data from: https://blog.nus.edu.sg/sshsphphg/

## Rtools REQUIRED
### Manual install, run .exe, restart R session: 
Need Rtools for R 3.5.1 to compile GenABEL from source
https://cran.r-project.org/bin/windows/Rtools/history.html


Setting up environment

```{r setup, include=FALSE}
if (!require("knitr", quietly = TRUE)){
    install.packages("knitr", version = '1.23')
}
knitr::opts_chunk$set(echo = TRUE)

## CHANGE WORKNG DIR
working_dir = '~/../Desktop/creigs_ta/GWAS/GWAStutorial-master/'
knitr::opts_knit$set(root.dir = working_dir)
```

```{r}
setwd(working_dir)
if (!require("GenABEL.data", quietly = TRUE)) {
    install.packages("GenABEL.data", repos="http://R-Forge.R-project.org", version = '1.0.0')
}

if (!require("GenABEL", quietly = TRUE)) {
  # download source
  url <- "https://cran.r-project.org/src/contrib/Archive/GenABEL/GenABEL_1.8-0.tar.gz"
  curl_command <- paste("curl", "-o", url)
  response <- system(curl_command, intern = TRUE)
  cat(response, sep = "\n")
  # build from source
  install.packages("GenABEL_1.8-0.tar.gz", repos = NULL)
}

if (!require("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")
}

if (!require("snpStats", quietly = TRUE)){
  BiocManager::install("snpStats")
}
  
if (!require("SNPRelate", quietly = TRUE)){
  BiocManager::install("SNPRelate")
}
  
if (!require("doParallel", quietly = TRUE)){
  install.packages("doParallel")
}

if (!require("dplyr", quietly = TRUE)){
  install.packages("dplyr")
}

if (!require("plyr", quietly = TRUE)){
  install.packages("plyr")
}
```

# Start tutorial

Load snps
```{r}
setwd(working_dir)

if(!require("snpStats")) {
      stop("snpStats is required for this script. Please install it on your system.")
}
load("conversionTable.RData")

pathM <- paste("public/Genomics/108Malay_2527458snps", c(".bed", ".bim", ".fam"), sep = "")
SNP_M <- read.plink(pathM[1], pathM[2], pathM[3])

pathI <- paste("public/Genomics/105Indian_2527458snps", c(".bed", ".bim", ".fam"), sep = "")
SNP_I <- read.plink(pathI[1], pathI[2], pathI[3])

pathC <- paste("public/Genomics/110Chinese_2527458snps", c(".bed", ".bim", ".fam"), sep = "")
SNP_C <- read.plink(pathC[1], pathC[2], pathC[3])

# Ensure == number of markers across the three populations
if(ncol(SNP_C$genotypes) != ncol(SNP_I$genotypes)){
        stop("Different number of columns in input files detected. This is not allowed.")
}
if(ncol(SNP_I$genotypes) != ncol(SNP_M$genotypes)){
        stop("Different number of columns in input files detected. This is not allowed.")
}

# Merge the three SNP datasets
SNP <- SNP_M
SNP$genotypes <- rbind(SNP_M$genotypes, SNP_I$genotypes, SNP_C$genotypes)
colnames(SNP$map) <- c("chr", "SNP", "gen.dist", "position", "A1", "A2") # same for all three
SNP$fam<- rbind(SNP_M$fam, SNP_I$fam, SNP_C$fam)

# Rename SNPs present in the conversion table into rs IDs
mappedSNPs <- intersect(SNP$map$SNP, names(conversionTable))
newIDs <- conversionTable[match(SNP$map$SNP[SNP$map$SNP %in% mappedSNPs], names(conversionTable))]
SNP$map$SNP[rownames(SNP$map) %in% mappedSNPs] <- newIDs

# Load lipid datasets & match SNP-Lipidomics samples
lipidsMalay <- read.delim("public/Lipidomic/117Malay_282lipids.txt", row.names = 1)
lipidsIndian <- read.delim("public/Lipidomic/120Indian_282lipids.txt", row.names = 1)
lipidsChinese <- read.delim("public/Lipidomic/122Chinese_282lipids.txt", row.names = 1)

all(Reduce(intersect, list(colnames(lipidsMalay),
                           colnames(lipidsIndian),
                           colnames(lipidsChinese))) == colnames(lipidsMalay)) # TRUE
lip <- rbind(lipidsMalay, lipidsIndian, lipidsChinese)

# Country
country <- sapply(list(SNP_M, SNP_I, SNP_C), function(k){
        nrow(k$genotypes)
})
origin <- data.frame(sample.id = rownames(SNP$genotypes),
                     Country = factor(rep(c("M", "I", "C"), country)))

matchingSamples <- intersect(rownames(lip), rownames(SNP$genotypes))
SNP$genotypes <- SNP$genotypes[matchingSamples,]
lip <- lip[matchingSamples,]
origin <- origin[match(matchingSamples, origin$sample.id),]
# Combine SNP and Lipidomics
genData <- list(SNP = SNP$genotype, MAP = SNP$map, LIP = lip)

# Write processed omics and GDS
save(genData, origin, file = "PhenoGenoMap.RData")
write.plink("convertGDS", snps = SNP$genotypes)

# Clear memory
# rm(list = ls())
```

```{r}
setwd(working_dir)

for(pkg in c("snpStats", "doParallel", "SNPRelate", "GenABEL")){
      if(!require(pkg, character.only = T)) {
            stop("At least one pckg is required for this script. Please install it on your system.")
      }
}

source("GWASfunction.R")
load("PhenoGenoMap.RData")

# Use SNP call rate of 100%, MAF of 0.1 (very stringent)
maf <- 0.1
callRate <- 1
SNPstats <- col.summary(genData$SNP)

maf_call <- with(SNPstats, MAF > maf & Call.rate == callRate)
genData$SNP <- genData$SNP[,maf_call]
genData$MAP <- genData$MAP[maf_call,]
SNPstats <- SNPstats[maf_call,]

# Sample call rate & heterozygosity
callMat <- !is.na(genData$SNP)
Sampstats <- row.summary(genData$SNP)
hetExp <- callMat %*% (2 * SNPstats$MAF * (1 - SNPstats$MAF)) # Hardy-Weinberg heterozygosity (expected)
hetObs <- with(Sampstats, Heterozygosity * (ncol(genData$SNP)) * Call.rate)
Sampstats$hetF <- 1-(hetObs/hetExp)
# Use sample call rate of 100%, het threshold of 0.1 (very stringent)
het <- 0.1 # Set cutoff for inbreeding coefficient;
het_call <- with(Sampstats, abs(hetF) < het & Call.rate == 1)
genData$SNP <- genData$SNP[het_call,]
genData$LIP <- genData$LIP[het_call,]

# LD and kinship coeff
ld <- .2
kin <- .1
snpgdsBED2GDS(bed.fn = "convertGDS.bed", bim.fn = "convertGDS.bim",
              fam.fn = "convertGDS.fam", out.gdsfn = "GDS",
              cvt.chr = "char")
genofile <- snpgdsOpen("GDS", readonly = F)
gds.ids <- read.gdsn(index.gdsn(genofile,  "sample.id"))
gds.ids <- sub("-1", "", gds.ids)
add.gdsn(genofile, "sample.id", gds.ids, replace = T)
geno.sample.ids <- rownames(genData$SNP)
# First filter for LD
snpSUB <- snpgdsLDpruning(genofile, ld.threshold = ld,
                          sample.id = geno.sample.ids,
                          snp.id = colnames(genData$SNP),
                          num.thread = 8)
snpset.ibd <- unlist(snpSUB, use.names = F)
# And now filter for MoM
ibd <- snpgdsIBDMoM(genofile, kinship = T,
                    sample.id = geno.sample.ids,
                    snp.id = snpset.ibd,
                    num.thread = 8)
ibdcoef <- snpgdsIBDSelection(ibd)
ibdcoef <- ibdcoef[ibdcoef$kinship >= kin,]

# Filter samples out
related.samples <- NULL
while (nrow(ibdcoef) > 0) {
      # count the number of occurrences of each and take the top one
      sample.counts <- sort(table(c(ibdcoef$ID1, ibdcoef$ID2)), decreasing = T)
      rm.sample <- names(sample.counts)[1]
      cat("Removing sample", rm.sample, "too closely related to",
          sample.counts[1], "other samples.\n")
      
      # remove from ibdcoef and add to list
      ibdcoef <- ibdcoef[ibdcoef$ID1 != rm.sample & ibdcoef$ID2 != rm.sample,]
      related.samples <- c(as.character(rm.sample), related.samples)
}
genData$SNP <- genData$SNP[!(rownames(genData$SNP) %in% related.samples),]
genData$LIP <- genData$LIP[!(rownames(genData$LIP) %in% related.samples),]
```

```{r}
setwd(working_dir)

# PCA
set.seed(100)
pca <- snpgdsPCA(genofile, sample.id = geno.sample.ids, 
                 snp.id = snpset.ibd, num.thread = 1)
pctab <- data.frame(sample.id = pca$sample.id,
                    PC1 = pca$eigenvect[,1],
                    PC2 = pca$eigenvect[,2],
                    stringsAsFactors = F)

# Subset and/or reorder origin accordingly
origin <- origin[match(pca$sample.id, origin$sample.id),]

pcaCol <- rep(rgb(0,0,0,.3), length(pca$sample.id)) # Set black for chinese
pcaCol[origin$Country == "I"] <- rgb(1,0,0,.3) # red for indian
pcaCol[origin$Country == "M"] <- rgb(0,.7,0,.3) # green for malay

png("PCApopulation.png", width = 500, height = 500)
plot(pctab$PC1, pctab$PC2, xlab = "PC1", ylab = "PC2", col = pcaCol, pch = 16)
abline(h = 0, v = 0, lty = 2, col = "grey")
legend("top", legend = c("Chinese", "Indian", "Malay"), col = 1:3, pch = 16, bty = "n")
dev.off()

# Choose trait for association analysis, use colnames(genData$LIP) for listing
# NOTE: Ignore the first column of genData$LIP (gender)
target <- "Cholesterol"

phenodata <- data.frame("id" = rownames(genData$LIP),
                        "phenotype" = scale(genData$LIP[,target]), stringsAsFactors = F)

# Conduct GWAS (will take a while)
start <- Sys.time()
GWAA(genodata = genData$SNP, phenodata = phenodata, filename = paste(target, ".txt", sep = ""))
Sys.time() - start # benchmark

# Manhattan plot
GWASout <- read.table(paste(target, ".txt", sep = ""), header = T, colClasses = c("character", rep("numeric",4)))
GWASout$type <- rep("typed", nrow(GWASout))
GWASout$Neg_logP <- -log10(GWASout$p.value)
GWASout <- merge(GWASout, genData$MAP[,c("SNP", "chr", "position")])
GWASout <- GWASout[order(GWASout$Neg_logP, decreasing = T),]

png(paste(target, ".png", sep = ""), height = 500,width = 1000)
GWAS_Manhattan(GWASout)
dev.off()

# QQ plot using GenABEL estlambda function
png(paste(target, "_QQplot.png", sep = ""), width = 500, height = 500)
lambda <- estlambda(GWASout$t.value**2, plot = T, method = "median")
dev.off()

writeLines(capture.output(sessionInfo()), "sessionInfo")
```
```{r}
saveRDS(genData, 'genData.rds')
saveRDS(GWASout, 'GWASout.rds')
saveRDS(pca, 'pca.rds')
saveRDS(SNPstats,'SNPstats.rds')
saveRDS(origin, 'origin.rds')
saveRDS(snpSUB ,'snpSUB.rds')
```


